### PR TITLE
Change lists to always be empty, never null.

### PR DESCRIPTION
**Pull request by JakeWharton**
Note: this suffers from #250

### CODE REVIEW

File: wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java [hunk 1]
```diff
@@ -210,9 +210,9 @@ private TypeSpec toTypeSpec(WireMessage type) {
     for (WireField field : type.fieldsAndOneOfFields()) {
       TypeName fieldType = fieldType(field);
 
-      if (field.type().isScalar()
-          || javaGenerator.isEnum(field.type())
-          || field.isRepeated()) {
+      if ((field.type().isScalar() || javaGenerator.isEnum(field.type()))
```
Comment (swankjesse) [inline wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java]: behavior change ?

Comment (JakeWharton) [inline wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java]: This change removes the emission of the default empty list constant which is no longer required. Scalars and enums still get defaults.

File: wire-runtime/src/main/java/com/squareup/wire/Message.java [hunk 2]
```diff
@@ -319,13 +312,16 @@ protected IllegalStateException missingRequiredFields(Object... args) {
...
      *
      * @return the incoming list.
      * @throws NullPointerException if a null element is present in the list.
      */
     protected static <T> List<T> checkForNulls(List<T> elements) {
-      if (elements != null && !elements.isEmpty()) {
+      if (elements == null) {
+        return Collections.emptyList();
+      }
+      if (!elements.isEmpty()) {
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/Message.java]: unnecessary?

**Comment by swankjesse** [comment]
LGTM

File: wire-runtime/src/test/java/com/google/protobuf/DescriptorProto.java [hunk 3]
```diff
@@ -139,15 +129,15 @@ public int hashCode() {
 
     public String doc;
 
-    public List<FieldDescriptorProto> field;
+    public List<FieldDescriptorProto> field = Collections.emptyList();
```
Comment (danrice-square) [inline wire-runtime/src/test/java/com/google/protobuf/DescriptorProto.java]: Can we emit an `import static` for `emptyList`?

File: wire-runtime/src/test/java/com/squareup/wire/WireTest.java [hunk 4]
```diff
@@ -324,4 +326,11 @@ public void testBadEnum() throws IOException {
     byte[] newData = result.toByteArray();
     assertTrue(Arrays.equals(data, newData));
   }
+
+  @Test public void missingRepeatedAndPackagedFieldsBecomesEmptyList() throws IOException {
```
Comment (danrice-square) [inline wire-runtime/src/test/java/com/squareup/wire/WireTest.java]: PackedFields

File: wire-runtime/src/test/java/com/squareup/wire/WireTest.java [hunk 5]
```diff
@@ -324,4 +326,11 @@ public void testBadEnum() throws IOException {
...
+    byte[] bytes = new byte[0];
+    RepeatedAndPacked allTypes = new Wire().parseFrom(bytes, RepeatedAndPacked.class);
```
Comment (danrice-square) [inline wire-runtime/src/test/java/com/squareup/wire/WireTest.java]: Change variable name `allTypes` to something better.

File: wire-runtime/src/test/java/com/google/protobuf/DescriptorProto.java (hunk 3 above)
Comment (JakeWharton) [inline wire-runtime/src/test/java/com/google/protobuf/DescriptorProto.java]: Unfortunately not at this time (https://github.com/square/javapoet/issues/73).

File: wire-runtime/src/test/java/com/squareup/wire/protos/RepeatedAndPacked.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    public Builder(RepeatedAndPacked message) {
+      super(message);
+      if (message == null) return;
+      this.rep_int32 = copyOf(message.rep_int32);
+      this.pack_int32 = copyOf(message.pack_int32);
+    }
+
+    public Builder rep_int32(List<Integer> rep_int32) {
+      this.rep_int32 = checkForNulls(rep_int32);
```
Comment (danrice-square) [inline wire-runtime/src/test/java/com/squareup/wire/protos/RepeatedAndPacked.java]: Let's find a better name than `checkForNulls`

**Comment by danrice-square** [comment]
LGTM with a few comments.

File: wire-runtime/src/test/java/com/squareup/wire/protos/RepeatedAndPacked.java (hunk 6 above)
Comment (JakeWharton) [inline wire-runtime/src/test/java/com/squareup/wire/protos/RepeatedAndPacked.java]: I chose `canonicalizeList`. Not married to it, but couldn't come up with anything better at the moment.

**Pull request merged**